### DISCUSSION TITLE
Issues with multi-behavioral camera + miniscope bonsai workflow

### DISCUSSION THREAD

**Discussion by spencerbyers15** [bonsai-rx/bonsai#1106] [Q&A] [answered]
Hello,

I am a new bonsai user trying to create a workflow that records video and timestamps from 4 behavioral cameras (3 FLIR and 1 Basler- all set to 60 fps with 'FMP4' codec) and 1 miniscope (generally set to 10 fps and being with 'DIB ' codec). I’ve run into a few problems with my bonsai workflow so far-

(1)    My workflow is saving many short videos from each camera source, using a TakeUntil and Timer loop which only stops when I click the STOP button in bonsai. I’ve noticed that there are **frames dropped between the writing of one video to the beginning of the next video** for each camera. Is this a common issue when writing so many high frame rate videos or does this stem from the looping I am performing on each image source? 
(2)    **Sometimes bonsai saves a timestamp csv with no corresponding video** although the timestamp csv indicates that frames should have been written. This only ever occurs for the last video file collected in a single session, and it does not happen every time I start a workflow. 
(3)    I would like to **write a video every 1000 frames captured** from each of these cameras, rather than the solution that I came up with where I use a TakeUntil and a Timer node to get ~1000 frames per video written. I’ve found a post on this topic here (https://groups.google.com/g/bonsai-users/c/Ia5HdQBMI4o/m/QYHOf0sWAwAJ) but I can’t find the operator ElementCountWindow as was mentioned in the answer in that post.
(4)    Finally (and this is a nitpicky one) I would like to be able to set a **path prefix** for all of the videos and csv timestamps to be written into a subfolder path that I can set before recording. As my workflow is now, I would have to alter filename properties for the VideoWriter and CsvWriter for all 5 camera streams to get them to all go into the same subdirectory. If I was recording long trials this would be okay but I am generally recording 1-2 minute videos and this would be a lot of paths to set that often. 

I’ve attached a png of my workflow and can send the .bonsai file too if that's useful. 

Thanks in advance for taking the time to help me with these issues!

Best,
Spencer Byers

![5_camera_bonsai](https://user-images.githubusercontent.com/42975884/199113062-4f60f48b-ab61-43cf-affc-4f2ea7144ed4.PNG)

**Comment by bruno-f-cruz** [comment, marked as answer]
1) You want to try to avoid restarting the camera source. Try something like this:
![image](https://user-images.githubusercontent.com/7049351/199192037-8f3df2bf-ace8-49f3-945f-8fa58687a48b.png)

2) This is weird, how are you closing the workflow? I have some changes to the code just in case

3) You can simply use "Take(1000)" instead of your TakeUntil.
[Example.zip](https://github.com/bonsai-rx/bonsai/files/9908220/Example.zip)

4) check the attached workflow for a way this could be done. You still need to increment the index of the file but that you can use the property of the writer nodes, if needed

**Comment by kapilsaxena33** [comment]
How do we control the frame rate of the flir camera. I set up in VideoWriter node the frame rate as 30, but I get around 67?

**Reply by bruno-f-cruz** [reply]
VideoWriter determines the encoding FPS of the video not the camera, they are completely independent. To set the frame rate of the camera, I would check the device's manual. There is likely a setting in their software (SpinView) or you can trigger it with an external TTL signal.
